docs/guide.md

```diff
@@ -0,0 +1,286 @@
+developers to specify their own encoding schemes.  This enables
+the SDK to be used as the framwork for constructing already
+specified cryptocurrency state machines, for instance Ethereum. 
+
+When initializing an application, a developer must specify a
+`TxDecoder` function which determines how an arbitrary byte array
+should be unmarshalled into a `Tx`: 
+
+```golang
+type TxDecoder func(txBytes []byte) (Tx, error)
+```
+
+In `Basecoin`, we use the Tendermint wire format and the
+`go-wire` library for encoding and decoding all message types.
+The `go-wire` library has the nice property that it can unmarshal
+into interface types, but it requires the relevant types to be
+registered ahead of type. Registration happens on a `Codec`
+object, so as not to taint the global name space.
+
+For instance, in `Basecoin`, we wish to register the `SendMsg`
+and `IssueMsg` types:
+
+```golang
+cdc.RegisterInterface((*sdk.Msg)(nil), nil)
+cdc.RegisterConcrete(bank.SendMsg{}, "cosmos-sdk/SendMsg", nil)
+cdc.RegisterConcrete(bank.IssueMsg{}, "cosmos-sdk/IssueMsg", nil)
+```
+
+Note how each concrete type is given a name - these name
+determines the types unique "prefix bytes" during encoding.  A
+registered type will always use the same prefix-bytes, regardless
+of what interface it is satisfying.  For more details, see the
+[go-wire documentation]().
+
+
+## MultiStore
+
+### MultiStore is like a filesystem
+### Mounting an IAVLStore
+
+```
+TODO:
+- IAVLStore: Fast balanced dynamic Merkle store.
+  - supports iteration.
+- MultiStore: multiple Merkle tree backends in a single store 
+  - allows using Ethereum Patricia Trie and Tendermint IAVL in same app
+- Provide caching for intermediate state during execution of blocks and transactions (including for iteration)
+- Historical state pruning and snapshotting.
+- Query proofs (existence, absence, range, etc.) on current and retained historical state.
+```
+
+## Context
+
+The SDK uses a `Context` to propogate common information across
+functions. The `Context` is modelled off of the Golang
+`context.Context` object, which has become ubiquitous in
+networking middleware and routing applications as a means to
+easily propogate request context through handler functions.
+
+The main information stored in the `Context` includes the
+application MultiStore (see below), the last block header, and
+the transaction bytes. Effectively, the context contains all data
+that may be necessary for processing a transaction.
+
+Many methods on SDK objects receive a context as the first
+argument. 
+
+## Handler
+
+Transaction processing in the SDK is defined through `Handler`
+functions:
+
+```golang
+type Handler func(ctx Context, tx Tx) Result
+```
+
+A handler takes a context and a transaction and returns a result.
+All information necessary for processing a transaction should be
+available in the context.
+
+While the context holds the entire application store, a
```

Comment:
A description of how the capability keys prevent unauthorised access to some parts of the store would be good I think.